File: project_euler/problem_118/solution.cpp
#include <bits/stdc++.h>
using namespace std;

// Miller-Rabin primality test for numbers up to ~10^9
bool is_prime(long long n) {
    if (n < 2) return false;
    if (n < 4) return true;
    if (n % 2 == 0 || n % 3 == 0) return false;
    for (long long i = 5; i * i <= n; i += 6) {
        if (n % i == 0 || n % (i + 2) == 0) return false;
    }
    return true;
}

int digits[9] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
// For each bitmask of digits, store all primes formable from those digits
map<int, vector<long long>> mask_primes;

void gen_perms(int mask, vector<int>& ds, int idx, long long num, vector<long long>& primes) {
    if (idx == (int)ds.size()) {
        if (is_prime(num)) primes.push_back(num);
        return;
    }
    for (int i = idx; i < (int)ds.size(); i++) {
        swap(ds[i], ds[idx]);
        gen_perms(mask, ds, idx + 1, num * 10 + ds[idx], primes);
        swap(ds[i], ds[idx]);
    }
}

void precompute() {
    // Enumerate all non-empty subsets of {1..9} (bitmask over 9 bits)
    for (int mask = 1; mask < 512; mask++) {
        vector<int> ds;
        int dsum = 0;
        for (int i = 0; i < 9; i++) {
            if (mask & (1 << i)) {
                ds.push_back(i + 1);
                dsum += (i + 1);
            }
        }
        // Pruning: if digit sum divisible by 3 and subset has >1 digit or the single digit is not 3
        // Actually, just compute all primes and filter
        vector<long long> primes;
        gen_perms(mask, ds, 0, 0, primes);
        // Remove duplicates and sort
        sort(primes.begin(), primes.end());
        primes.erase(unique(primes.begin(), primes.end()), primes.end());
        if (!primes.empty()) {
            mask_primes[mask] = primes;
        }
    }
}

int answer = 0;

void solve(int remaining_mask, long long last_prime) {
    if (remaining_mask == 0) {
        answer++;
        return;
    }
    // Enumerate all non-empty subsets of remaining_mask
    for (int sub = remaining_mask; sub > 0; sub = (sub - 1) & remaining_mask) {
        auto it = mask_primes.find(sub);
        if (it == mask_primes.end()) continue;
        for (long long p : it->second) {
            if (p > last_prime) {
                solve(remaining_mask ^ sub, p);
            }
        }
    }
}

int main() {
    precompute();
    solve(511, 0); // 511 = (1<<9)-1, all 9 digits available
    cout << answer << endl;
    return 0;
}

File: project_euler/problem_118/solution.py
"""
Problem 118: Pandigital Prime Sets

Find the number of distinct sets of primes that use each digit 1-9 exactly once.
"""

from itertools import permutations

def isprime(n):
    """Deterministic primality test for n up to ~10^9."""
    if n < 2:
        return False
    if n < 4:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

def solve():
    # Precompute: for each bitmask of digits {1..9}, find all primes
    mask_primes = {}

    for mask in range(1, 512):
        digits = []
        for i in range(9):
            if mask & (1 << i):
                digits.append(i + 1)

        primes = set()
        for perm in permutations(digits):
            num = int(''.join(map(str, perm)))
            if isprime(num):
                primes.add(num)

        if primes:
            mask_primes[mask] = sorted(primes)

    # DFS: partition all 9 digits into sets forming primes in increasing order
    count = 0

    def dfs(remaining, last_prime):
        nonlocal count
        if remaining == 0:
            count += 1
            return

        # Enumerate non-empty subsets of remaining
        sub = remaining
        while sub > 0:
            if sub in mask_primes:
                for p in mask_primes[sub]:
                    if p > last_prime:
                        dfs(remaining ^ sub, p)
            sub = (sub - 1) & remaining

    dfs(511, 0)
    return count

answer = solve()
assert answer == 44680
print(answer)
